frontend/expr_inferrer.py

```diff
@@ -375,33 +376,74 @@ def _get_args_types(args, context, instance, solver):
     return args_types
 
 
-def _infer_annotated_function_call(args_types, solver, annotations):
-    result_type = solver.new_z3_const("call")
-    solver.annotation_resolver.add_annotated_function_axioms(args_types, solver, annotations, result_type)
+def _infer_annotated_function_call(args_types, solver, annotations, result_type):
+    return solver.annotation_resolver.get_annotated_function_axioms(args_types, solver, annotations, result_type)
 
-    return result_type
+
+def _infer_builtin_method(args_types, solver, context, result_type, method_name):
+    """Get the axioms of built-in methods inference"""
+    possible_methods = context.get_matching_methods(method_name)
+    method_axioms = []
+    for method in possible_methods:
+        cur_method_axioms = _infer_annotated_function_call(args_types, solver, method, result_type)
+        if cur_method_axioms is not None:
+            method_axioms.append(cur_method_axioms)
+    return method_axioms
 
 
 def infer_func_call(node, context, solver):
     """Infer the type of a function call, and unify the call types with the function parameters"""
     instance = None
+    result_type = solver.new_z3_const("call")
+    call_axioms = []
     if isinstance(node.func, ast.Attribute):
-        called, instance = infer(node.func, context, solver, True)
+        # Add axioms for built-in methods
+        instance = infer(node.func.value, context, solver)
         args_types = _get_args_types(node.args, context, instance, solver)
+        call_axioms += _infer_builtin_method(args_types, solver, context, result_type, node.func.attr)
     else:
         args_types = _get_args_types(node.args, context, instance, solver)
         if isinstance(node.func, ast.Name) and context.has_annotated_func(node.func.id):
-            return _infer_annotated_function_call(args_types, solver, context.get_annotated_func(node.func.id))
+            # Add axioms for built-in functions
+            annotated_func = context.get_annotated_func(node.func.id)
+            func_axioms = _infer_annotated_function_call(args_types, solver, annotated_func, result_type)
+            if func_axioms is not None:
+                call_axioms.append(func_axioms)
+
+    try:
+        # Add normal call inference axioms
         called = infer(node.func, context, solver)
-
-    result_type = solver.new_z3_const("call")
-
-    solver.add(axioms.call(called, args_types, result_type, solver.z3_types),
+        call_axioms += axioms.call(called, args_types, result_type, solver.z3_types)
+    except NameError as e:
+        # function name wasn't found in the context
+        if not call_axioms:
+            # no match for the function name in the context or in the built-ins
+            raise e
```

Comment:
Yeah I guess the message fits pretty well, but we will still get a stack trace that looks exactly the way an accidental stack trace that occurs because of a programming error in the inference would look like, and there is very little in terms of information for the use (where is the reference to the undefined name?). 
I think we should have one kind of exception that we throw when the input program is somehow invalid, catch it somewhere, and translate it to a proper error message (preferably with line information or something like that). The user shouldn't see stack traces for normal use cases like "untypable input program", and it should be easy to distinguish whether something in the inference has gone wrong or not.